src/sqlfluff/rules/L064.py

```diff
@@ -0,0 +1,262 @@
+"""Implementation of Rule L064."""
+
+import sys
+from typing import Optional
+
+import regex
+
+if sys.version_info < (3, 8):
+    from typing_extensions import Final
+else:
+    from typing import Final
+
+from sqlfluff.core.parser.segments.raw import CodeSegment
+from sqlfluff.core.rules.base import BaseRule, LintFix, LintResult, RuleContext
+from sqlfluff.core.rules.doc_decorators import (
+    document_configuration,
+    document_fix_compatible,
+)
+
+QUOTES_MAPPING: Final = {
+    "single_quotes": {
+        "common_name": "single quotes",
+        "preferred_quote_char": "'",
+        "alternate_quote_char": '"',
+    },
+    "double_quotes": {
+        "common_name": "double quotes",
+        "preferred_quote_char": '"',
+        "alternate_quote_char": "'",
+    },
+}
+# BigQuery string prefix characters.
+STRING_PREFIX_CHARS: Final = "rbRB"
+
+
+@document_configuration
+@document_fix_compatible
+class Rule_L064(BaseRule):
+    r"""Consistent usage of preferred quotes for STRING datatype.
+
+    Some databases allow single quotes as well as double quotes to express a STRING
+    literal. Prefer one type of quotes as specified in rule setting. Falling back to
+    alternate quotes to reduce escapes.
+
+    Dollar quoted raw strings are excluded from this rule, as they are mostly used for
+    literal UDF Body definitions.
+
+    .. note::
+       This rule can be dangerous. If accidentally enabled for dialects that do not
+       support single *and* double quotes automated fixes can potentially break working
+       SQL code.
+
+       This rule is disabled by default for dialetcs that don't allow double quotes for
+       STRING datatype (e.g. Postgres). It can be enabled with the
+       ``force_enable = True`` flag.
+
+    **Anti-pattern**
+
+    .. code-block:: sql
+       :force:
+
+        select
+            "abc",
+            'abc',
+            "\"",
+            "abc" = 'abc'
+        from foo
+
+    **Best practice**
+
+    Ensure all STRING literals use preferred quotes, unless escaping can be reduced by
+    using alternate quotes.
+
+    .. code-block:: sql
+       :force:
+
+        select
+            "abc",
+            "abc",
+            '"',
+            "abc" = "abc"
+        from foo
+
+    """
+
+    config_keywords = ["preferred_string_quotes", "force_enable"]
+    _dialects_with_double_quoted_strings = [
+        "bigquery",
+        "hive",
+        "mysql",
+        "sparksql",
+    ]
+
+    def _eval(self, context: RuleContext) -> Optional[LintResult]:
+        # Config type hints
+        self.preferred_string_quotes: str
+        self.force_enable: bool
+
+        if (
+            context.dialect.name not in self._dialects_with_double_quoted_strings
+            and not self.force_enable
+        ):
+            return LintResult(memory=context.memory)
+
+        # Only care about quoted literal segments.
+        if not context.segment.name == "quoted_literal":
+            return None
```

Comment:
Why only `quoted_literal` and not also `quoted_identifiers`?

Is it because in some dialects quoted_identifiers are no in single/double quotes (e.g. square brackets in TSQL, back ticks in BigQuery)? If so, I get that, but maybe want to make it more clear in the documentation as I (incorrectly!) thought it was literals and identifiers.

Also think BigQuery can be in-scope if it's just literals: https://cloud.google.com/bigquery/docs/reference/standard-sql/lexical
And maybe also MySQL unless the default of ANSI_QUOTE is used?: https://dev.mysql.com/doc/refman/8.0/en/string-literals.html

Reply:
My knowledge in other SQL dialects other than BigQuery and a bit of Postgres is actually not that large. From what I've seen rules for identifiers are pretty wild. However, for this rule to work we need a notion of preferred and alternate quotes. I don't even know if there is any dialect that has similar rule set for identifiers?

Even if there is I would prefer to avoid this for now and potentially add this when the use-case arises?

I agree that the documentation should be clear on that. Which part made you beliebe that identifiers were also covered by this rule? I tried to use the terminology of "STRING datatype" in rule documentation. But I'd be open for another (more clear) wording.

Reply:
> Which part made you beliebe that identifiers were also covered by this rule? 

I just assumed. ass-you-and-me applies :-)

I thin we should use "quoted literals" over "STRING datatype". That's more consistent with what we use in other rules, for example: https://docs.sqlfluff.com/en/stable/rules.html#sqlfluff.core.rules.Rule_L048

Also suggest we explicitly call out that quoted identifiers are out of scope.

Made some suggested wording changes in latest review.

Reply:
See you accepted those suggestions!

Only outstanding question on this comment thread then is whether we bring BigQuery and MySQL into scope then?

Reply:
What do you mean by into scope here?

Reply:
You’re excluding those dialects but for quoted literals I don’t think you need to exclude them? I agree for quotes identifiers they have special rules but since that’s not in scope we don’t need to worry about them.

Reply:
I'm including them 😂 Currently, this rule is enabled for all dialects in this list https://github.com/sqlfluff/sqlfluff/pull/3118/files#diff-c74eb3c9d616e01ddba16d42924a3a06bee0828c413408b642f6255e8c3a53d7R69-R74
so `bigquery`, `hive `, `mysql` and `sparksql`.

Reply:
Ah no you’re NOT excluding - you’re including! Ignore me.

However, since you’re using an inclusion pattern I wonder if we can tone down the warning slightly?

Reply:
Yeah, maybe even remove the warning entirely? It's a left over from the approach where it was disabled by default. I think it should be clear to users that if they put `force_enable = True` that they are leaving territory the rule was intended for. WDYT?

Reply:
I say leave it in, but removing first sentence. Made some more suggestions.